libs/sdk-core/src/persist/swap_taproot.rs

```diff
@@ -0,0 +1,433 @@
+                );
+            }
+
+            let current_swap = result
+                .get_mut(&address)
+                .expect("just inserted key is not in hashmap");
+            self.map_tx_info(current_swap, row)?;
+            self.map_refund(current_swap, row)?;
+        }
+
+        Ok(result.into_values().collect())
+    }
+
+    fn map_tx_info(
+        &self,
+        swap: &mut FullTaprootSwapData,
+        row: &Row<'_>,
+    ) -> Result<(), PersistError> {
+        let tx_id: Option<String> = row.get("tx_id")?;
+        let tx_id = match tx_id {
+            Some(tx_id) => tx_id,
+            None => return Ok(()),
+        };
+        let output_index: u32 = row.get("output_index")?;
+
+        if swap
+            .outputs
+            .iter()
+            .any(|o| o.tx_id == tx_id && o.output_index == output_index)
+        {
+            return Ok(());
+        }
+
+        let mut output = TaprootSwapOutput {
+            address: swap.swap.address.clone(),
+            amount_sat: row.get("amount_sat")?,
+            tx_id: tx_id.clone(),
+            output_index,
+            confirmed_at_height: row.get("confirmed_at_height")?,
+            block_hash: row.get("block_hash")?,
+            spend: None,
+        };
+
+        let spending_tx_id: Option<String> = row.get("spending_tx_id")?;
+        if let Some(spending_tx_id) = spending_tx_id {
+            output.spend = Some(TaprootSwapSpend {
+                tx_id,
+                output_index,
+                spending_tx_id,
+                spending_input_index: row.get("spending_input_index")?,
+                block_hash: row.get("spend_block_hash")?,
+                confirmed_at_height: row.get("spend_confirmed_at_height")?,
+            })
+        }
+
+        swap.outputs.push(output);
+        Ok(())
+    }
+
+    fn map_refund(
+        &self,
+        swap: &mut FullTaprootSwapData,
+        row: &Row<'_>,
+    ) -> Result<(), PersistError> {
+        let refund_tx_id: Option<String> = row.get("refund_tx_id")?;
+        let refund_tx_id = match refund_tx_id {
+            Some(refund_tx_id) => refund_tx_id,
+            None => return Ok(()),
+        };
+
+        if swap.refund_transactions.contains(&refund_tx_id) {
+            return Ok(());
+        }
+
+        swap.refund_transactions.push(refund_tx_id);
+        Ok(())
+    }
+
+    pub(crate) fn list_unused_taproot_swaps(
+        &self,
+    ) -> Result<Vec<FullTaprootSwapData>, PersistError> {
+        self.get_full_taproot_swaps(" WHERE t.tx_id IS NULL", &[])
+    }
+
+    pub(crate) fn list_taproot_swaps(&self) -> Result<Vec<FullTaprootSwapData>, PersistError> {
+        self.get_full_taproot_swaps("", &[])
+    }
+
+    pub(crate) fn set_taproot_swap_bolt11(
+        &self,
+        address: &str,
+        bolt11: &str,
+    ) -> Result<(), PersistError> {
+        let mut conn = self.get_connection()?;
+        let tx = conn.transaction_with_behavior(TransactionBehavior::Immediate)?;
```

Comment:
Don't need the immediate write behaviour

Reply:
Done in 3c17ebcb92a40b934386df2ac48679b1214398e3